### PR TITLE
Batch RL

### PR DESCRIPTION
**Pull request by galleibo-intel**
A WIP PR for batch reinforcement learning support.

### CODE REVIEW

File: rl_coach/graph_managers/batch_rl_graph_manager.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        # TODO handle unsupported use-cases for batch RL. i.e. on-policy algorithms,
+        #  distributed setting (either local or multi-node)
+
+        # initialize the network parameters from the global network
+        self.sync()
+
+        # heatup
+        if self.env_params is not None:
+            self.heatup(self.heatup_steps)
```
Comment (galnov) [inline rl_coach/graph_managers/batch_rl_graph_manager.py]: There shouldn't be any interaction with the environment, so what does heatup mean in this context?

File: rl_coach/agents/agent_interface.py [hunk 2]
```diff
@@ -173,3 +173,12 @@ def handle_episode_ended(self) -> None:
         :return: None
         """
         raise NotImplementedError("")
+    
+    def run_ope(self) -> None:
```
Comment (galnov) [inline rl_coach/agents/agent_interface.py]: Suggest to change to a meaningful (even if longer) name such as run_offline_policy_eval(...)

File: rl_coach/off_policy_evaluators/ope_manager.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        :param network_keys: The network keys used for feeding the neural networks.
+        :param temperature: The Softmax temperature to use for estimating a probability from Q values.
+        :return:
+        """
+        # IPS
+        all_reward_model_rewards, all_policy_probs, all_old_policy_probs = [], [], []
+        all_v_values_reward_model_based, all_v_values_q_model_based, all_rewards, all_actions = [], [], [], []
+
+        for i in range(int(len(dataset_as_transitions) / batch_size) + 1):
+            batch = dataset_as_transitions[i * batch_size: (i + 1) * batch_size]
```
Comment (galnov) [inline rl_coach/off_policy_evaluators/ope_manager.py]: Is there a guarantee that number of transitions is a multiply of batch size?

Comment (galleibo-intel) [reply rl_coach/off_policy_evaluators/ope_manager.py]: No, and that's the reason for the `+ 1`. the last sampled batch might be smaller than batch_size. If the last batch is of size 0 (i.e. transitions number *is* a multiply of batch size), then nothing gets appended to each of the lists.

File: rl_coach/off_policy_evaluators/ope_manager.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        all_reward_model_rewards = np.concatenate(all_reward_model_rewards, axis=0)
+        all_policy_probs = np.concatenate(all_policy_probs, axis=0)
+        all_v_values_reward_model_based = np.concatenate(all_v_values_reward_model_based, axis=0)
+        all_rewards = np.concatenate(all_rewards, axis=0)
+        all_actions = np.concatenate(all_actions, axis=0)
+        all_old_policy_probs = np.concatenate(all_old_policy_probs, axis=0)
+
+        # generate model probabilities
+        new_policy_prob = np.max(all_policy_probs, axis=1)
+        rho_all_dataset = new_policy_prob / all_old_policy_probs
```
Comment (galnov) [inline rl_coach/off_policy_evaluators/ope_manager.py]: Is it possible that a new_policy_prob is for a different action than the old_policy_prob, so rho is not calculated correctly?
new_policy_prob is the max among all actions as predicted by the q_network, and old_policy_prob is for the action that was actually taken. Maybe the predicted action is different than the one that was actually taken?

File: rl_coach/presets/CartPole_DQN_BatchRL.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+# Test #
+########
+preset_validation_params = PresetValidationParameters()
+preset_validation_params.test = True
+preset_validation_params.min_reward_threshold = 150
+preset_validation_params.max_episodes_to_achieve_reward = 250
+
+graph_manager = BatchRLGraphManager(agent_params=agent_params, env_params=env_params,
+                                    schedule_params=schedule_params,
+                                    vis_params=VisualizationParameters(dump_signals_to_csv_every_x_episodes=1),
```
Comment (galnov) [inline rl_coach/presets/CartPole_DQN_BatchRL.py]: Is dashboard supposed to work with this preset?
I'm getting some weird looking signals (all zeros except for last episode).

Comment (galleibo-intel) [reply rl_coach/presets/CartPole_DQN_BatchRL.py]: Yes, it does. Since this is BatchRL, the x-axis has to be changed to Training Iterations (instead of episodes). I will have Dashboard change the default view in the case of BatchRL to Training Iterations.

File: rl_coach/agents/agent_interface.py (hunk 2 above)
Comment (galleibo-intel) [reply rl_coach/agents/agent_interface.py]: done.

File: rl_coach/graph_managers/batch_rl_graph_manager.py (hunk 1 above)
Comment (galleibo-intel) [reply rl_coach/graph_managers/batch_rl_graph_manager.py]: This is for cases where we  want to do batch RL, based on an existing environment simulator, which we collect the dataset from. This is useful mainly for tutorial and debug purposes.

File: rl_coach/off_policy_evaluators/ope_manager.py (hunk 4 above)
Comment (galleibo-intel) [reply rl_coach/off_policy_evaluators/ope_manager.py]: Yes, this is a bug. Thanks!!!

File: rl_coach/presets/CartPole_DQN_BatchRL.py (hunk 5 above)
Comment (galleibo-intel) [reply rl_coach/presets/CartPole_DQN_BatchRL.py]: done. dashboard is changing the default x-axis type according to the graph type.

File: rl_coach/agents/value_optimization_agent.py [hunk 6]
```diff
@@ -32,6 +33,7 @@ def __init__(self, agent_parameters, parent: Union['LevelManager', 'CompositeAge
         super().__init__(agent_parameters, parent)
         self.q_values = self.register_signal("Q")
         self.q_value_for_action = {}
+        self.first = True
```
Comment (galnov) [inline rl_coach/agents/value_optimization_agent.py]: What is "first" for and when is it used?

File: rl_coach/graph_managers/basic_rl_graph_manager.py [hunk 7]
```diff
@@ -31,17 +32,24 @@ class BasicRLGraphManager(GraphManager):
...
+                 preset_validation_params: PresetValidationParameters = PresetValidationParameters(),
+                 name='simple_rl_graph'):
+        super().__init__(name, schedule_params, vis_params)
         self.agent_params = agent_params
         self.env_params = env_params
         self.preset_validation_params = preset_validation_params
-
         self.agent_params.visualization = vis_params
+
         if self.agent_params.input_filter is None:
```
Comment (galnov) [inline rl_coach/graph_managers/basic_rl_graph_manager.py]: Need some documentation on why this change is needed

File: rl_coach/agents/value_optimization_agent.py (hunk 6 above)
Comment (galleibo-intel) [reply rl_coach/agents/value_optimization_agent.py]: It is not, probably some leftover. Will remove. Thanks!

File: rl_coach/graph_managers/basic_rl_graph_manager.py (hunk 7 above)
Comment (galleibo-intel) [reply rl_coach/graph_managers/basic_rl_graph_manager.py]: It is needed in cases where there is no environment. Today the use-cases are batch-rl and imitation learning, and there might be more in the future. In those case, since there is not environment, there is no where to get a default filter from. So using a default no filter.
